File: app/core/src/test/java/com/fsck/k9/notification/RemoveNotificationResultTest.java
package com.fsck.k9.notification;


import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;


public class RemoveNotificationResultTest {
    private static final int NOTIFICATION_ID = 23;


    private NotificationHolder notificationHolder;


    @Before
    public void setUp() throws Exception {
        notificationHolder = new NotificationHolder(NOTIFICATION_ID, null);
    }

    @Test
    public void createNotification_shouldCancelNotification_shouldReturnTrue() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.createNotification(notificationHolder);

        assertTrue(result.shouldCreateNotification());
    }

    @Test
    public void createNotification_getNotificationId_shouldReturnNotificationId() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.createNotification(notificationHolder);

        assertEquals(NOTIFICATION_ID, result.getNotificationId());
    }

    @Test
    public void createNotification_isUnknownNotification_shouldReturnFalse() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.createNotification(notificationHolder);

        assertFalse(result.isUnknownNotification());
    }

    @Test
    public void createNotification_getNotificationHolder_shouldReturnNotificationHolder() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.createNotification(notificationHolder);

        assertEquals(notificationHolder, result.getNotificationHolder());
    }

    @Test
    public void cancelNotification_shouldCancelNotification_shouldReturnFalse() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.cancelNotification(NOTIFICATION_ID);

        assertFalse(result.shouldCreateNotification());
    }

    @Test
    public void cancelNotification_getNotificationId_shouldReturnNotificationId() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.cancelNotification(NOTIFICATION_ID);

        assertEquals(NOTIFICATION_ID, result.getNotificationId());
    }

    @Test
    public void cancelNotification_isUnknownNotification_shouldReturnFalse() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.cancelNotification(NOTIFICATION_ID);

        assertFalse(result.isUnknownNotification());
    }

    @Test(expected = IllegalStateException.class)
    public void cancelNotification_getNotificationHolder_shouldThrowException() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.cancelNotification(NOTIFICATION_ID);

        result.getNotificationHolder();
    }

    @Test
    public void unknownNotification_shouldCancelNotification_shouldReturnFalse() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.unknownNotification();

        assertFalse(result.shouldCreateNotification());
    }

    @Test(expected = IllegalStateException.class)
    public void unknownNotification_getNotificationId_shouldThrowException() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.unknownNotification();

        result.getNotificationId();
    }

    @Test
    public void unknownNotification_isUnknownNotification_shouldReturnTrue() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.unknownNotification();

        assertTrue(result.isUnknownNotification());
    }

    @Test(expected = IllegalStateException.class)
    public void unknownNotification_getNotificationHolder_shouldThrowException() throws Exception {
        RemoveNotificationResult result = RemoveNotificationResult.unknownNotification();

        result.getNotificationHolder();
    }
}

File: app/core/src/test/java/com/fsck/k9/notification/RemoveNotificationResultTest.kt
package com.fsck.k9.notification

import com.fsck.k9.controller.MessageReference
import com.fsck.k9.notification.RemoveNotificationResult.Companion.cancelNotification
import com.fsck.k9.notification.RemoveNotificationResult.Companion.createNotification
import com.fsck.k9.notification.RemoveNotificationResult.Companion.unknownNotification
import com.google.common.truth.Truth.assertThat
import org.junit.Test

private const val NOTIFICATION_ID = 23

class RemoveNotificationResultTest {
    private val notificationHolder = NotificationHolder(
        notificationId = NOTIFICATION_ID,
        content = NotificationContent(
            messageReference = MessageReference("irrelevant", 1, "irrelevant"),
            sender = "irrelevant",
            subject = "irrelevant",
            preview = "irrelevant",
            summary = "irrelevant"
        )
    )

    @Test
    fun createNotification_shouldCancelNotification_shouldReturnTrue() {
        val result = createNotification(notificationHolder)

        assertThat(result.shouldCreateNotification).isTrue()
    }

    @Test
    fun createNotification_getNotificationId_shouldReturnNotificationId() {
        val result = createNotification(notificationHolder)

        assertThat(result.notificationId).isEqualTo(NOTIFICATION_ID)
    }

    @Test
    fun createNotification_isUnknownNotification_shouldReturnFalse() {
        val result = createNotification(notificationHolder)

        assertThat(result.isUnknownNotification).isFalse()
    }

    @Test
    fun createNotification_getNotificationHolder_shouldReturnNotificationHolder() {
        val result = createNotification(notificationHolder)

        assertThat(result.notificationHolder).isEqualTo(notificationHolder)
    }

    @Test
    fun cancelNotification_shouldCancelNotification_shouldReturnFalse() {
        val result = cancelNotification(NOTIFICATION_ID)

        assertThat(result.shouldCreateNotification).isFalse()
    }

    @Test
    fun cancelNotification_getNotificationId_shouldReturnNotificationId() {
        val result = cancelNotification(NOTIFICATION_ID)

        assertThat(result.notificationId).isEqualTo(NOTIFICATION_ID)
    }

    @Test
    fun cancelNotification_isUnknownNotification_shouldReturnFalse() {
        val result = cancelNotification(NOTIFICATION_ID)

        assertThat(result.isUnknownNotification).isFalse()
    }

    @Test(expected = IllegalStateException::class)
    fun cancelNotification_getNotificationHolder_shouldThrowException() {
        val result = cancelNotification(NOTIFICATION_ID)

        result.notificationHolder
    }

    @Test
    fun unknownNotification_shouldCancelNotification_shouldReturnFalse() {
        val result = unknownNotification()

        assertThat(result.shouldCreateNotification).isFalse()
    }

    @Test(expected = IllegalStateException::class)
    fun unknownNotification_getNotificationId_shouldThrowException() {
        val result = unknownNotification()

        result.notificationId
    }

    @Test
    fun unknownNotification_isUnknownNotification_shouldReturnTrue() {
        val result = unknownNotification()

        assertThat(result.isUnknownNotification).isTrue()
    }

    @Test(expected = IllegalStateException::class)
    fun unknownNotification_getNotificationHolder_shouldThrowException() {
        val result = unknownNotification()

        result.notificationHolder
    }
}
